Fix constant folding bug with sym size tensor

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #118411

When there was a constant folded SymInt which was used to construct a then constant folding tensor, we had previously used tried to use the sympy symbol which would error (should take in SymInt not symbol). 

Fix by recording the observed size during constant folding.


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

eellison committed:
Fix constant folding bug with sym size tensor

[ghstack-poisoned]

eellison commented on torch/_inductor/fx_passes/joint_graph.py:
```diff
@@ -238,7 +242,7 @@ def constant_fold_uniform_value(gm: torch.fx.GraphModule):
             new_node = graph.call_function(
                 aten.full.default,
-                args=(list(fake_tensor.shape), value),
+                args=(list(node_replacements_shapes[node]), value),
```
We're taking the Size from the runtime tensor which will not SymInts. I can make it `List[int]` instead for better typing.

ezyang replied:
This still seems suspicious? Suppose previously the representation of the tensor in the graph had dynamic shape (e.g., was expressed in terms of s0). After this change, you will specialize the tensor. To make matters worse, it's not clear to me that you have installed the right guard!

What I was *expecting* to see was to rewrite the transformation to preserve the input node structure for sizes, rewriting only other parts of the constructor?

eellison committed:
Update on "Fix constant folding bug with sym size tensor"


When there was a constant folded SymInt which was used to construct a then constant folding tensor, we had previously used tried to use the sympy symbol which would error (should take in SymInt not symbol). 

Fix by recording the observed size during constant folding.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

eellison committed:
Update on "Fix constant folding bug with sym size tensor"


When there was a constant folded SymInt which was used to construct a then constant folding tensor, we had previously used tried to use the sympy symbol which would error (should take in SymInt not symbol). 

Fix by recording the observed size during constant folding.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

ezyang reviewed (approved):
I'd like a comment elaborating on what we discussed in WC

eellison committed:
Update on "Fix constant folding bug with sym size tensor"


When there was a constant folded SymInt which was used to construct a then constant folding tensor, we had previously used tried to use the sympy symbol which would error (should take in SymInt not symbol). 

Fix by recording the observed size during constant folding.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

eellison committed:
Update on "Fix constant folding bug with sym size tensor"


When there was a constant folded SymInt which was used to construct a then constant folding tensor, we had previously used tried to use the sympy symbol which would error (should take in SymInt not symbol). 

Fix by recording the observed size during constant folding.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

eellison committed:
Update on "Fix constant folding bug with sym size tensor"


When there was a constant folded SymInt which was used to construct a then constant folding tensor, we had previously used tried to use the sympy symbol which would error (should take in SymInt not symbol). 

Fix by recording the observed size during constant folding.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

eellison committed:
Update on "Fix constant folding bug with sym size tensor"


When there was a constant folded SymInt which was used to construct a then constant folding tensor, we had previously used tried to use the sympy symbol which would error (should take in SymInt not symbol). 

Fix by recording the observed size during constant folding.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

eellison commented:
@pytorchbot merge
